```{r paper_setup, cache=FALSE, include=FALSE}
library(knitcitations)
# create bibtex entries
write.bibtex(c(knitr=citation('knitr'),
               knitcitations=citation('knitcitations'), plyr=citation('plyr'),
               ggplot2=citation('ggplot2'), ape=citation('ape'),
               directlabels=citation('directlabels'),
               plyr=citation('plyr'), foreach=citation('foreach')),
             file='citations.bib')
bib = read.bibtex('citations.bib')
```

# PrimerTree: Visualizing polymerase chain reaction primer product alignments for metagenomic studies #
# Abstract #
## Summary ##
Design and screening of primers is one of the most critical steps in conducting
a polymerase chain reaction (PCR) analysis.  It is important for the primers
to amplify correctly and to be both sensitive and specific to the samples of
interest.  In metagenomic studies, it is also important to amplify a
representative sample of a given population.  Prior tools exist for designing
primers based on a template and for identifying potential target sequences in
a database.  However no tool to systematically analyze and visualize
the sensitivity and specificity of PCR products currently exists. PrimerTree is
a package which given a set of primers finds all known targets and then generates
and visualizes a phylogenetic tree of the primer products annotated with the
taxonomic information of the samples.

## Availability ##
PrimerTree is an R package released under the GPL-2 license and is available through
[CRAN](http://cran.r-project.org/web/packages/primerTree/index.html).  Source
code and developmental versions are available at <http://www.github.com/jimhester/primerTree>.

## Contact ##
[hesterj@ccf.org](mailto:hesterj@ccf.org)

# Introduction #
Designing primers for PCR is a common task widely used in biomedical research,
forensic science and metagenomics `r citep(c('10.1371/journal.pbio.0050082'))`.
There are a number of important factors to consider in primer design. First one
must consider the chemical properties of the primers themselves, including
length, melting temperature, GC content, secondary structure and the likelihood
of primer dimer formation.  In addition, the primer must be specific to the
target of interest and must not produce off target products.  Metagenomic
studies must design primers that amplify a representative subset of the species
of interest, while not amplifying unrelated species.  Primer products also must
carry enough differences to be identifiable and distinct after sequencing.

Researchers without a strong computational background often carry out primer
design. Additionally, the results should be interpretable without a much prior
phylogenetic experience. Therefore, the tools used to design the
primers and perform the analysis must be easy to use and understand.  They must
also be available on a wide variety of systems.

While there exists simple online tools for primer design `r citep('10.1093/nar/gks596')`,
and primer specificity checking `r citep(c(primer_BLAST='10.1186/1471-2105-13-134'))`
, there are not any simple tools to systematically compare primer products.
The PrimerTree R package addresses this lack by generating a phylogenetic tree
of the primer products along with their taxonomic classification given a pair
of primers.  This gives the researcher a good idea of the scope and sensitivity
of the primers in the environment.   Because it is an R package, it runs on a
wide variety of hardware and the entire analysis can run in only two
commands, making it accessible to a large audience.

# Implementation #
PrimerTree performs five main functions in its analysis,
primer search, taxonomic identification of the found sequences, multiple
sequence alignment of the primer products, construction of a phylogenetic tree,
and visualization of the tree by taxonomic rank.

## Primer search ##
PrimerTree utilizes the primer search implemented in Primer-BLAST
`r citep('primer_BLAST')` by directly querying the National Center for
Biotechnology Information (NCBI) Primer-BLAST search page. This allows one to
use all the options available on the NCBI site. By default, PrimerTree
searches the NCBI (nt) nucleotide database to provide the most complete dataset for
metagenomic studies.  Alternative NCBI databases can be chosen through the
function options if necessary.  The primer alignment results are retrieved and
parsed, and the primer products retrieved from NCBI's servers using
the NCBI E-utilties.
`r citep('http://www.ncbi.nlm.nih.gov/books/NBK25500/')`.

## Taxonomic identification ##
NCBI's taxonomy database `r citep('http://www.ncbi.nlm.nih.gov/books/NBK21100/')`
provides a full set of taxonomic information for all identified samples in
their databases.  Primer-BLAST retrieves this information for the samples
obtained in the search using NCBI E-Utilities and the results are
stored for later use.

## Multiple sequence alignment ##
Next, Clustal Omega constructs the alignment of primer product sequences `r citep('10.1038/msb.2011.75')`.
Clustal Omega is the latest installment of the widely used ClustalW2 multiple
alignment software, and is available as source
code as well as precompiled binaries for Mac OSX, Windows, Linux and FreeBSD.
This allows easy installation of Clustal Omega by any user of PrimerTree.  This
alignment is then used to construct a phylogenetic tree.

## Phylogenetic tree construction ##
The final step in the PrimerTree analysis is to construct a phylogenetic tree of
the multiple alignment.  A distance matrix of the sequences is calculated and a
neighbor-joining tree constructed using the ape package `r citep('10.1093/bioinformatics/btg412')`.
The tree and the previously retrieved taxonomic ranks can then be visualized.

## Tree visualization by taxonomic rank ##
PrimerTree visualizes the tree, coloring the main taxonomic ranks using
the ggplot2 package, then adds group names to the colored ranks using the
directlabels package `r citep(c(bib[['ggplot2']], bib[['directlabels']]))`.
Fig. 1 shows an example of commonly used universal primers from the chloroplast trnL gene.
`r citep('10.1093/nar/gkl938')`
```{r figure_1, dev=c('png', 'tiff'), fig.width=12, fig.height=12, dpi=300, fig.cap='Fig. 1 - PrimerTree results from universal primers for the chloroplast trnL gene.  Note class is omitted from the plot because most of the sequences did not have a class annotation', echo=F, message=F, warning=F}
library(primerTree)
plot(trnL, ranks=c('kingdom', 'phylum', 'order', 'family', 'genus', 'species'), size=4)
```

# Runtime characteristics #
The trnL primer pair is a simple case having no degenerate primers.  When using a
single thread primer alignment takes about 120 seconds, retrieval of taxonomic
information takes about 10 seconds, retrieval of primer product sequences takes
about 100 seconds, multiple alignment takes about 3 seconds, and tree layout
and printing takes about 2 seconds.  The entire analysis completes in less
than 240 seconds.  The average runtime for the analysis varies greatly based
the specificity of the primers as well as the search parameters chosen, as well
as the current load on the NCBI servers and the internet connection, therefore
more accurate runtime estimates are not possible.

Highly degenerate primer pairs result in a large number of possible
permutations to search with Primer-BLAST, which can increase the runtime
considerably.  In an effort to limit maximum runtime in this situation, PrimerTree
randomly samples only a portion of the total primer permutations.  This
provides a representative sample for most cases.  Changing the number of
sampled permutations or turning off sampling completely is also possible using
function options.

PrimerTree uses the plyr package extensively, so has full support for any of the
parallel backends compatible with the foreach package
`r citep(c(bib[['plyr']], bib[['foreach']]))`.  In particular, parallel retrieval of the primer
sequences from NCBI speeds up the total runtime considerably. Note that
parallel queries to Primer-BLAST are queued by NCBI's servers and only
processed once there is free compute time.

# Conclusions #
Designing primers for any experiment can be a daunting challenge, especially so
if it is a metagenomic study.  We have developed the PrimerTree package to help
visualize the sensitivity and specificity of the primers chosen. PrimerTree is
an R package available on a wide variety of platforms and is very simple to use
and install.  PrimerTree is released under an open-source license and can be
downloaded from
[http://github.com/jimhester/primerTree](http://github.com/jimhester/PrimerTree),
which also provides additional documentation.

## Conflict of Interest ##
None declared.

# References #
```{r paper_bibliography, results='asis', echo=F}
bibliography(style='markdown')
```
