```{r paper_setup, cache=FALSE, include=FALSE}
library(knitcitations)
```

# PrimerTree: Visualizing polymerase chain reaction primer target alignments for metagenomics #
# Abstract #
## Summary ##
Design and screening of primers is one of the most critical steps in conducting
any polymerase chain reaction (PCR) analysis.  It is important for the primers
to amplify correctly and to be both sensitive and specific to the samples of
interest.  In metagenomic studies it is also important to amplify a
representative sample of a given population.  Prior tools exist for designing
primers based on a template and for identifying potential target sequences in
sequence databases.  However no tool currently exists to systematically analyze
the diversity of the PCR products, so the scope and precision of the primers
can be assessed.  PrimerTree is a package which given a set of 
primers finds all known targets, and generates and visualizes a phylogenetic
tree of the primer products annotated with the taxonomic information of the
samples.

## Availability ##
PrimerTree is available as a R package under the GPL-2 license available through
[CRAN](http://cran.r-project.org/web/packages/primerTree/index.html).  Source
code and developmental versions available at <http://www.github.com/jimhester/primerTree>.

## Contact ##
[hesterj@ccf.org](mailto:hesterj@ccf.org)

# Introduction #
Designing primers for PCR is a common task widely used in biomedical research,
forensic science and recently, metagenomics `r citet(c('10.1371/journal.pbio.0050082'))`.
There are a number of important factors to consider in designing primers which
will correctly amplify the target of interest.  First chemical properties of
the primers themselves must be considered, including length, melting
temperature, GC content, secondary structure and the likelihood of primer dimer
formation.  In addition, the primer must be specific to the target of interest
and must not produce off target products.  Metagenomic studies must design
primers which amplify a representative subset of the species of interest, while
not amplifying those species which are unrelated.  If the primer products are
to be sequenced, the primer products must carry enough differences to be
identifiable and distinct.

Primer design is often carried out by researchers without a strong computational
background.  Therefore the tools used to design the primers and perform the analysis
must be easy to use and available on a wide variety of systems.

While there exists simple online tools for primer design `r citet('10.1186/1471-2105-13-134')`,
and primer specificity checking `r citet('10.1385/1-59259-192-2:365')`, there
are not any currently available easy to use tools to systematically compare
primer products.  PrimerTree, an R package addresses this need by taking a
primer pair, generates a phylogenetic tree of the primer products along with
their taxonomic classification.  This gives the researcher a good idea of the
scope and sensitivity of the primers.  Because it is an R package, it runs on a
wide variety of hardware and the entire analysis can be run in only two
commands, making it accessible to a large audience.

# Implementation #
TODO

# Results and Discussion #
TODO

# Conclusions #
TODO

```{r paper_bibliography, results='asis', echo=F}
bibliography(style='markdown')
```
