<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Visually Assessing the Specificity and Informativeness of Primer
Pairs</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Visually Assessing the Specificity and Informativeness of Primer
Pairs
<img class="toplogo" src="../../../doc/html/logo.jpg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;primerTree&rsquo; version 1.0.3</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="primerTree.html">primerTree-package</a></td>
<td>'primerTree' Visually Assessing the Specificity and Informativeness of Primer Pairs</td></tr>
<tr><td style="width: 25%;"><a href="accession2gi.html">accession2gi</a></td>
<td>Maps a genbank accession to a nuclotide database gi.</td></tr>
<tr><td style="width: 25%;"><a href="bryophytes_trnL.html">bryophytes_trnL</a></td>
<td>PrimerTree results for the bryophyte trnL primers</td></tr>
<tr><td style="width: 25%;"><a href="calc_rank_dist_ave.html">calc_rank_dist_ave</a></td>
<td>calc_rank_dist_ave Summarize pairwise differences.</td></tr>
<tr><td style="width: 25%;"><a href="clustalo.html">clustalo</a></td>
<td>Multiple sequence alignment with clustal omega</td></tr>
<tr><td style="width: 25%;"><a href="filter_seqs.html">filter_seqs</a></td>
<td>Filter out sequences retrieved by search_primer_pair() that are either too short or too long. The alignment and tree will be recalculated after removing unwanted reads.</td></tr>
<tr><td style="width: 25%;"><a href="filter_seqs.html">filter_seqs.primerTree</a></td>
<td>Filter out sequences retrieved by search_primer_pair() that are either too short or too long. The alignment and tree will be recalculated after removing unwanted reads.</td></tr>
<tr><td style="width: 25%;"><a href="get_sequence.html">get_sequence</a></td>
<td>Retrieves a fasta sequence from NCBI nucleotide database.</td></tr>
<tr><td style="width: 25%;"><a href="get_sequences.html">get_sequences</a></td>
<td>Retrieves fasta sequences from NCBI nucleotide database.</td></tr>
<tr><td style="width: 25%;"><a href="get_taxonomy.html">get_taxonomy</a></td>
<td>Retrieve the taxonomy information from NCBI for a set of nucleotide gis.</td></tr>
<tr><td style="width: 25%;"><a href="gi2taxid.html">gi2taxid</a></td>
<td>Maps a nucleotide database gi to a taxonomy database taxId</td></tr>
<tr><td style="width: 25%;"><a href="identify.primerTree_plot.html">identify.primerTree_plot</a></td>
<td>identify the point closest to the mouse click only works on single ranks</td></tr>
<tr><td style="width: 25%;"><a href="layout_tree_ape.html">layout_tree_ape</a></td>
<td>layout a tree using ape, return an object to be plotted by 'plot_tree'</td></tr>
<tr><td style="width: 25%;"><a href="mammals_16S.html">mammals_16S</a></td>
<td>PrimerTree results for the mammalian 16S primers</td></tr>
<tr><td style="width: 25%;"><a href="plot.primerTree.html">plot.primerTree</a></td>
<td>plot function for a primerTree object, calls plot_tree_ranks</td></tr>
<tr><td style="width: 25%;"><a href="plot_tree.html">plot_tree</a></td>
<td>plots a tree, optionally with colored and labeled points by taxonomic rank</td></tr>
<tr><td style="width: 25%;"><a href="plot_tree_ranks.html">plot_tree_ranks</a></td>
<td>plots a tree along with a series of taxonomic ranks</td></tr>
<tr><td style="width: 25%;"><a href="primerTree.html">primerTree</a></td>
<td>'primerTree' Visually Assessing the Specificity and Informativeness of Primer Pairs</td></tr>
<tr><td style="width: 25%;"><a href="primer_search.html">primer_search</a></td>
<td>Query a pair of primers using ncbi's Primer-BLAST, if primers contain iupac</td></tr>
<tr><td style="width: 25%;"><a href="search_primer_pair.html">search_primer_pair</a></td>
<td>Automatic primer searching Search a given primer pair, retrieving the alignment results, their product sequences, the taxonomic information for the sequences, a multiple alignment of the products</td></tr>
<tr><td style="width: 25%;"><a href="seq_lengths.html">seq_lengths</a></td>
<td>Get a summary of sequence lengths from a primerTree object</td></tr>
<tr><td style="width: 25%;"><a href="seq_lengths.primerTree.html">seq_lengths.primerTree</a></td>
<td>Method for primerTree objects</td></tr>
<tr><td style="width: 25%;"><a href="summary.primerTree.html">summary.primerTree</a></td>
<td>Summarize a primerTree result, printing quantiles of sequence length and pairwise differences.</td></tr>
<tr><td style="width: 25%;"><a href="tree_from_alignment.html">tree_from_alignment</a></td>
<td>Construct a neighbor joining tree from a dna alignment</td></tr>
</table>
</body></html>
