<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Automatic primer searching Search a given primer pair,...</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for search_primer_pair {primerTree}"><tr><td>search_primer_pair {primerTree}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Automatic primer searching Search a given primer pair, retrieving the alignment
results, their product sequences, the taxonomic information for the sequences,
a multiple alignment of the products</h2>

<h3>Description</h3>

<p>Automatic primer searching Search a given primer pair, retrieving the alignment
results, their product sequences, the taxonomic information for the sequences,
a multiple alignment of the products
</p>


<h3>Usage</h3>

<pre>
search_primer_pair(forward, reverse, name = NULL, num_aligns = 500,
  num_permutations = 25, simplify = TRUE, clustal_options = list(),
  distance_options = list(model = "N", pairwise.deletion = T), ...,
  .parallel = FALSE, .progress = "none")
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>forward</code></td>
<td>
<p>forward primer to search by 5'-3' on plus strand</p>
</td></tr>
<tr valign="top"><td><code>reverse</code></td>
<td>
<p>reverse primer to search by 5'-3' on minus strand</p>
</td></tr>
<tr valign="top"><td><code>name</code></td>
<td>
<p>name to give to the primer pair</p>
</td></tr>
<tr valign="top"><td><code>num_aligns</code></td>
<td>
<p>number of alignment results to keep</p>
</td></tr>
<tr valign="top"><td><code>num_permutations</code></td>
<td>
<p>the number of primer permutations to search, if the degenerate bases
cause more than this number of permutations to exist, this number will be
sampled from all possible permutations.</p>
</td></tr>
<tr valign="top"><td><code>simplify</code></td>
<td>
<p>use simple names for primer hit results or complex</p>
</td></tr>
<tr valign="top"><td><code>clustal_options</code></td>
<td>
<p>a list of options to pass to clustal omega, see
<code>link{clustalo}</code> for a list of options</p>
</td></tr>
<tr valign="top"><td><code>distance_options</code></td>
<td>
<p>a list of options to pass to dist.dna, see
<code>link{dist.dna}</code> for a list of options</p>
</td></tr>
<tr valign="top"><td><code>...</code></td>
<td>
<p>additional arguments passed to Primer-Blast</p>
</td></tr>
<tr valign="top"><td><code>.parallel</code></td>
<td>
<p>if 'TRUE', perform in parallel, using parallel backend
provided by foreach</p>
</td></tr>
<tr valign="top"><td><code>.progress</code></td>
<td>
<p>name of the progress bar to use, see
<code><a href="../../plyr/html/create_progress_bar.html">create_progress_bar</a></code></p>
</td></tr>
</table>


<h3>Value</h3>

<p>A list with the following elements,
</p>
<table summary="R valueblock">
<tr valign="top"><td><code>name</code></td>
<td>
<p>name of the primer pair</p>
</td></tr>
<tr valign="top"><td><code>BLAST_result</code></td>
<td>
<p>html blast results from Primer-BLAST as
'a <code><a href="../../dismo/html/response.html">response</a></code></p>
</td></tr></table>
<p> object.
</p>
<table summary="R valueblock">
<tr valign="top"><td><code>taxonomy</code></td>
<td>
<p>taxonomy for the primer products from NCBI</p>
</td></tr>
<tr valign="top"><td><code>sequence</code></td>
<td>
<p>sequence of the primer products</p>
</td></tr>
<tr valign="top"><td><code>alignment</code></td>
<td>
<p>multiple alignment of the primer products</p>
</td></tr>
<tr valign="top"><td><code>tree</code></td>
<td>
<p>phylogenetic tree of the reconstructed from the
'multiple alignment</p>
</td></tr>
</table>


<h3>See Also</h3>

<p><code><a href="primer_search.html">primer_search</a></code>, <code><a href="clustalo.html">clustalo</a></code>
</p>


<h3>Examples</h3>

<pre>
## Not run: 
#simple search
mammals_16S = search_primer_pair(name='Mammals 16S',
 'CGGTTGGGGTGACCTCGGA', 'GCTGTTATCCCTAGGGTAACT')
#returning 1000 alignments, allow up to 3 mismatches in primer
mammals_16S = search_primer_pair(name='Mammals 16S',
 'CGGTTGGGGTGACCTCGGA', 'GCTGTTATCCCTAGGGTAACT',
 num_aligns=1000, total_primer_specificity_mismatch=3)

## End(Not run)
</pre>

<hr /><div style="text-align: center;">[Package <em>primerTree</em> version 1.0.3 <a href="00Index.html">Index</a>]</div>
</body></html>
